middleware/access.go

```diff
@@ -1,79 +1,137 @@
 package middleware
 
 import (
-	"context"
-	"fmt"
 	"net/http"
-	"time"
 
 	"github.com/0xsequence/quotacontrol/proto"
-	"github.com/go-chi/httprate"
-)
 
-const (
-	HeaderAccessKey = "X-Access-Key"
-	HeaderOrigin    = "Origin"
+	"github.com/go-chi/jwtauth/v5"
+	"github.com/lestrrat-go/jwx/v2/jwt"
 )
 
-// Client is the interface that wraps the basic FetchKeyQuota, GetUsage and SpendQuota methods.
-type Client interface {
-	IsEnabled() bool
-	FetchKeyQuota(ctx context.Context, accessKey, origin string, now time.Time) (*proto.AccessQuota, error)
-	FetchUsage(ctx context.Context, quota *proto.AccessQuota, now time.Time) (int64, error)
-	FetchUserPermission(ctx context.Context, projectID uint64, userID string, useCache bool) (*proto.UserPermission, *proto.ResourceAccess, error)
-	SpendQuota(ctx context.Context, quota *proto.AccessQuota, computeUnits int64, now time.Time) (bool, error)
-}
+// SetCredentials gets the credentials from the JWT or the Access key.
+// It uses JWT from `jwtauth.Verifier` to extract the project claim and sets it in the context as well.
+// When both are present, it checks project mismatch between the access key and the JWT token.
+func SetCredentials(ja *jwtauth.JWTAuth, accessKeyFuncs ...func(*http.Request) string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var (
+				ctx       = r.Context()
+				projectID = uint64(0)
+				accessKey = r.Header.Get(HeaderAccessKey)
+			)
```

Comment:
Please assign the values to variables right away with `:=` and without `var`
IMHO this way it is more readable.
```suggestion
		ctx := r.Context()
		accessKey := r.Header.Get(HeaderAccessKey)
```

And define `projectID` as in comment down below, when it is needed